operator/standalone/update.go

```diff
@@ -0,0 +1,386 @@
+func (u *UpdateStandalonePipeline) ensureHelmRelease(ctx context.Context) error {
+	chart := &u.config.Spec.HelmReleaseTemplate.Chart
+	helmValues, err := helmvalues.Marshal(u.helmValues)
+	if err != nil {
+		return err
+	}
+	u.helmRelease = &helmv1beta1.Release{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   u.instance.Status.HelmChart.DeploymentNamespace,
+			Labels: getCommonLabels(u.instance.Name),
+		},
+	}
+
+	updateRelease := func() error {
+		u.helmRelease.Spec = helmv1beta1.ReleaseSpec{
+			ForProvider: helmv1beta1.ReleaseParameters{
+				Chart:               helmv1beta1.ChartSpec{Repository: chart.Repository, Name: chart.Name, Version: chart.Version},
+				Namespace:           u.instance.Status.HelmChart.DeploymentNamespace,
+				SkipCreateNamespace: true,
+				SkipCRDs:            true,
+				Wait:                true,
+				ValuesSpec:          helmv1beta1.ValuesSpec{Values: helmValues},
+			},
+			ResourceSpec: crossplanev1.ResourceSpec{
+				ProviderConfigReference: &crossplanev1.Reference{Name: u.config.Spec.HelmProviderConfigReference},
+			},
+		}
+		return nil
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, u.client, u.helmRelease, updateRelease)
+	return err
+}
+
+func (u *UpdateStandalonePipeline) fetchOperatorConfig(ctx context.Context) error {
+	list := &v1alpha1.PostgresqlStandaloneOperatorConfigList{}
+	labels := client.MatchingLabels{
+		v1alpha1.PostgresqlMajorVersionLabelKey: u.instance.Spec.Parameters.MajorVersion.String(),
+	}
+	ns := client.InNamespace(u.operatorNamespace)
+	err := u.client.List(ctx, list, labels, ns)
+	if err != nil {
+		return err
+	}
+	if len(list.Items) == 0 {
+		return fmt.Errorf("no %s found with label '%s' in namespace '%s'", v1alpha1.PostgresqlStandaloneOperatorConfigKind, labels, ns)
+	}
+	if len(list.Items) > 1 {
+		return fmt.Errorf("multiple versions of %s found with label '%s' in namespace '%s'", v1alpha1.PostgresqlStandaloneOperatorConfigKind, labels, ns)
+	}
+	u.config = &list.Items[0]
+	return nil
+}
+
+func (u *UpdateStandalonePipeline) ensureConnectionSecret(ctx context.Context) error {
+	var updateSecretFunc func() error
+	if u.instance.Spec.Parameters.EnableSuperUser {
+		credentialsSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+			Name:      getCredentialSecretName(),
+			Namespace: u.instance.Status.HelmChart.DeploymentNamespace,
+		}}
+		err := getClientFromContext(ctx).Get(ctx, types.NamespacedName{Name: credentialsSecret.Name, Namespace: credentialsSecret.Namespace}, credentialsSecret)
+		if err != nil {
+			return err
+		}
+		updateSecretFunc = func() error {
+			u.connectionSecret.Data["POSTGRESQL_POSTGRES_PASSWORD"] = credentialsSecret.Data["postgres-password"]
+			return nil
+		}
+	} else {
+		updateSecretFunc = func() error {
+			delete(u.connectionSecret.Data, "POSTGRESQL_POSTGRES_PASSWORD")
+			return nil
+		}
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, u.client, u.connectionSecret, updateSecretFunc)
```

Comment:
This can be greatly simplified
```suggestion
	credentialsSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
		Name:      getCredentialSecretName(),
		Namespace: u.instance.Status.HelmChart.DeploymentNamespace,
	}}
	err := getClientFromContext(ctx).Get(ctx, client.ObjectKeyFromObject(credentialsSecret), credentialsSecret)
	if err != nil {
		return err
	}
	// the part above should be its own pipeline step

	_, err = controllerutil.CreateOrUpdate(ctx, getClientFromContext(ctx), u.connectionSecret, func() error {
		if u.instance.Spec.Parameters.EnableSuperUser {
			u.connectionSecret.Data["POSTGRESQL_POSTGRES_PASSWORD"] = credentialsSecret.Data["postgres-password"]
		} else {
			delete(u.connectionSecret.Data, "POSTGRESQL_POSTGRES_PASSWORD")
		}
		return nil
	})
	return err
```

Reply:
I already refactored this from another suggestion you made. Also here you always fetch the credential secret which is not the optimal solution.I moved the whole logic inside the mutating function and I'm fetching only when `u.instance.Spec.Parameters.EnableSuperUser `is true